lib/manageiq/messaging/kafka/common.rb

```diff
@@ -27,47 +32,55 @@ def queue_consumer
         end
 
         def raw_publish(commit, body, options)
-          producer.produce(encode_body(body), options)
+          producer.produce(encode_body(options[:headers], body), options)
           producer.deliver_messages if commit
-          logger.info("Published to topic(#{options[:topic]}), msg(#{payload_log(body[:data].inspect)})")
+          logger.info("Published to topic(#{options[:topic]}), msg(#{payload_log(body.inspect)})")
         end
 
         def queue_for_publish(options)
           body, kafka_opts = for_publish(options)
-          body[:message_type] = options[:message] if options[:message]
-          body[:class_name] = options[:class_name] if options[:class_name]
+          kafka_opts[:headers] = Hash(kafka_opts[:headers]).merge(:message_type => options[:message]) if options[:message]
+          kafka_opts[:headers] = Hash(kafka_opts[:headers]).merge(:class_name => options[:class_name]) if options[:class_name]
 
           [body, kafka_opts]
         end
 
         def topic_for_publish(options)
           body, kafka_opts = for_publish(options)
-          body[:event_type] = options[:event] if options[:event]
+          kafka_opts[:headers] = Hash(kafka_opts[:headers]).merge(:event_type => options[:event]) if options[:event]
 
           [body, kafka_opts]
         end
 
         def for_publish(options)
-          kafka_opts = {:topic => options[:service]}
-          kafka_opts[:partition_key] = options[:affinity] if options[:affinity]
+          kafka_opts = {:topic => address(options)}
+          kafka_opts[:partition_key] = options[:group_name] if options[:group_name]
+          kafka_opts[:headers] = {:sender => options[:sender]} if options[:sender]
```

Comment:
Hm maybe always initialize `kafka_opts[:headers]` to `{}` here so you don't have to worry about it being `nil` [here](https://github.com/ManageIQ/manageiq-messaging/pull/30/files#r222043593)?